File: sdk/typescript/tests/state/roomStateMachine.test.ts
import { describe, expect, test } from "bun:test";
import {
  HeartbeatTracker,
  NetplayInputFrameBatcher,
  PauseCoordinator,
  RuntimeTelemetryTracker,
  RoomStateMachine,
  StateHashReporter,
  type ServerMessage,
} from "../../src/index.ts";
import { roomView } from "../support/fixtures.ts";

describe("TypeScript netplay room state", () => {
  test("room joined stores reconnect ticket and assigned player", () => {
    const stateMachine = new RoomStateMachine();

    const state = stateMachine.apply({
      eventSeq: 1,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 1, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    expect(state.assignedPlayerIndex).toBe(0);
    expect(stateMachine.reconnectTokens.current()).toEqual({
      playerIndex: 0,
      resumeToken: "token",
      roomEpoch: 2,
    });
  });

  test("recovery resync updates state and keeps assignment", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 1,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 1, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    const state = stateMachine.apply({
      eventSeq: 10,
      room: roomView({
        eventSeq: 10,
        roomEpoch: 5,
        sessionEpoch: 9,
        status: "checkingCompatibility",
      }),
      roomEpoch: 5,
      sessionEpoch: 9,
      type: "recoveryResyncRequired",
    });

    expect(state.assignedPlayerIndex).toBe(0);
    expect(state.latestEventSeq).toBe(10);
    expect(state.room?.status).toBe("checkingCompatibility");
    expect(state.resync).toMatchObject({
      eventSeq: 10,
      phase: "requested",
      reason: "recovery",
      role: "host",
      roomEpoch: 5,
      sessionEpoch: 9,
      mustSendSnapshot: true,
    });
    expect(stateMachine.resync.shouldPauseEmulation()).toBe(true);
    expect(stateMachine.resync.shouldClearPredictionBuffers()).toBe(true);
    expect(stateMachine.reconnectTokens.current()?.roomEpoch).toBe(5);
  });

  test("state hash mismatch enters resync until session restarts", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 1,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 1, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    const mismatch = {
      frame: 120,
      hashes: [
        { playerIndex: 0, sha256: "a".repeat(64) },
        { playerIndex: 1, sha256: "b".repeat(64) },
      ],
      nearbyMatches: [],
    };
    const resyncing = stateMachine.apply({
      eventSeq: 11,
      mismatch,
      room: roomView({
        eventSeq: 11,
        roomEpoch: 2,
        sessionEpoch: 4,
        status: "checkingCompatibility",
      }),
      roomEpoch: 2,
      sessionEpoch: 4,
      type: "stateHashMismatch",
    });

    expect(resyncing.resync).toMatchObject({
      eventSeq: 11,
      mismatch,
      phase: "requested",
      reason: "stateHashMismatch",
      roomEpoch: 2,
      sessionEpoch: 4,
    });

    const started = stateMachine.apply({
      eventSeq: 12,
      room: roomView({
        eventSeq: 12,
        roomEpoch: 2,
        sessionEpoch: 4,
        status: "playing",
      }),
      roomEpoch: 2,
      sessionEpoch: 4,
      startFrame: 0,
      type: "startSession",
    });

    expect(started.resync).toBeNull();
  });

  test("resync coordinator exposes snapshot decisions", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 1,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 1, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });
    stateMachine.apply({
      eventSeq: 11,
      mismatch: {
        frame: 120,
        hashes: [
          { playerIndex: 0, sha256: "a".repeat(64) },
          { playerIndex: 1, sha256: "b".repeat(64) },
        ],
        nearbyMatches: [],
      },
      room: roomView({
        eventSeq: 11,
        roomEpoch: 2,
        sessionEpoch: 4,
        status: "checkingCompatibility",
      }),
      roomEpoch: 2,
      sessionEpoch: 4,
      type: "stateHashMismatch",
    });

    stateMachine.resync.markSnapshotNeeded(1_000);

    expect(stateMachine.resync.shouldSendHostSnapshot()).toBe(true);
    expect(stateMachine.resync.shouldWaitForSnapshot()).toBe(false);
  });

  test("reset clears room, reconnect, and pause state", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 1,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 1, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });
    stateMachine.apply({
      eventSeq: 2,
      pause: {
        acknowledgedPlayerIndexes: [],
        holders: [],
        pauseAtFrame: 50,
        pausedAtFrame: null,
        reason: "menu",
        requestedByPlayerIndex: 0,
        sequence: 1,
        state: "pausing",
      },
      room: roomView({ eventSeq: 2, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "sessionPauseScheduled",
    });

    stateMachine.reset();

    expect(stateMachine.state.room).toBeNull();
    expect(stateMachine.state.assignedPlayerIndex).toBeNull();
    expect(stateMachine.pause.currentPause).toBeNull();
    expect(stateMachine.resync.currentResync).toBeNull();
    expect(stateMachine.reconnectTokens.current()).toBeNull();
  });

  test("heartbeat tracker reports stale and recovery timeout", () => {
    const tracker = new HeartbeatTracker({
      recoveryAfterMs: 10_000,
      staleAfterMs: 5_000,
    });
    tracker.markAck(
      {
        eventSeq: 1,
        roomEpoch: 2,
        sessionEpoch: 3,
        type: "heartbeatAck",
      },
      1_000,
    );

    expect(tracker.health(1_500)).toBe("fresh");
    expect(tracker.health(6_000)).toBe("stale");
    expect(tracker.health(11_000)).toBe("recoveryTimedOut");
    expect(
      tracker.heartbeatMessage({
        latestEventSeq: 1,
        localFrame: 24,
        roomEpoch: 2,
        runtimeState: "playing",
        sessionEpoch: 3,
      }).runtimeState,
    ).toBe("playing");
  });

  test("heartbeat can consume runtime telemetry safely", () => {
    const tracker = new HeartbeatTracker();
    const telemetry = new RuntimeTelemetryTracker();

    telemetry.markLocalFrame(90);
    telemetry.recordRoundTrip(40);
    telemetry.recordRoundTrip(48);
    telemetry.recordStall();
    telemetry.recordCatchUpFrames(2);

    const heartbeat = tracker.heartbeatMessage({
      latestEventSeq: 4,
      roomEpoch: 2,
      runtimeState: "playing",
      sessionEpoch: 3,
      telemetry,
    });

    expect(heartbeat.localFrame).toBe(90);
    expect(heartbeat.network).toMatchObject({
      catchUpFrames: 2,
      roundTripMs: 48,
      stallCount: 1,
    });
    expect(telemetry.snapshot().network.stallCount).toBe(0);
  });

  test("state hash reporter normalizes hashes and deduplicates frames", () => {
    const reporter = new StateHashReporter({ reportEveryFrames: 30 });

    expect(reporter.shouldReport(0)).toBe(true);
    reporter.stateHashMessage({
      frame: 0,
      roomEpoch: 2,
      sessionEpoch: 3,
      sha256: "a".repeat(64),
    });
    expect(reporter.shouldReport(29)).toBe(false);
    expect(reporter.shouldReport(30)).toBe(true);
    expect(
      reporter.stateHashMessage({
        frame: 30,
        roomEpoch: 2,
        sessionEpoch: 3,
        sha256: "A".repeat(64),
      }).report.sha256,
    ).toBe("a".repeat(64));
    expect(reporter.shouldReport(30)).toBe(false);
    expect(reporter.shouldReport(59)).toBe(false);
    expect(reporter.shouldReport(61)).toBe(true);
  });

  test("pause coordinator creates and clears pause messages", () => {
    const pause = new PauseCoordinator(() => "request-1");
    pause.apply({
      acknowledgedPlayerIndexes: [],
      holders: [],
      pauseAtFrame: 120,
      pausedAtFrame: null,
      reason: "menu",
      requestedByPlayerIndex: 0,
      sequence: 3,
      state: "pausing",
    });

    expect(
      pause.pauseReached({
        pausedAtFrame: 121,
        roomEpoch: 4,
        sessionEpoch: 5,
      }).sequence,
    ).toBe(3);
    expect(
      pause.requestResume({
        reason: "menu",
        roomEpoch: 4,
        sessionEpoch: 5,
      }).requestId,
    ).toBe("request-1");

    pause.clear(3);
    expect(pause.currentPause).toBeNull();
  });

  test("relay errors are stored as close reasons", () => {
    const stateMachine = new RoomStateMachine();
    const message: ServerMessage = {
      code: "snapshotInvalid",
      message: "Snapshot payload is invalid.",
      type: "error",
    };

    const state = stateMachine.apply(message);

    expect(state.lastError).toEqual({
      code: "snapshotInvalid",
      kind: "relayError",
      message: "Snapshot payload is invalid.",
    });
  });

  test("input batcher rejects frames for the wrong player", () => {
    const batcher = new NetplayInputFrameBatcher({
      flush: () => {
        throw new Error("Unexpected input batch flush.");
      },
    });

    expect(() =>
      batcher.enqueue(
        {
          playerIndex: 0,
          roomEpoch: 2,
          sessionEpoch: 3,
        },
        {
          frame: 12,
          payload: [1],
          playerIndex: 1,
        },
      ),
    ).toThrow("Netplay input frame player does not match batch context.");
  });

  test("frame clock tracks server frame and peer read frame", () => {
    const stateMachine = new RoomStateMachine();

    stateMachine.apply({
      frame: {
        canonicalFrame: 20,
        frame: 18,
        roomEpoch: 1,
        sessionEpoch: 1,
      },
      type: "serverFrame",
    });
    stateMachine.apply({
      input: {
        frame: 16,
        payload: [1],
        playerIndex: 1,
      },
      roomEpoch: 1,
      sessionEpoch: 1,
      type: "inputFrame",
    });
    stateMachine.frameClock.markLocalFrame(80);

    expect(stateMachine.frameClock.snapshot()).toMatchObject({
      canonicalFrame: 20,
      peerReadFrame: 16,
      serverFrame: 18,
      stalled: true,
    });
  });

  test("stale epoch room messages are ignored", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 5,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 5, roomEpoch: 3, sessionEpoch: 4 }),
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    stateMachine.apply({
      eventSeq: 2,
      room: roomView({ eventSeq: 2, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomStateChanged",
    });

    expect(stateMachine.state.latestEventSeq).toBe(5);
    expect(stateMachine.state.roomEpoch).toBe(3);
  });

  test("stale same-epoch room messages are ignored", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 5,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 5, roomEpoch: 3, sessionEpoch: 4 }),
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    stateMachine.apply({
      eventSeq: 7,
      room: roomView({ eventSeq: 7, roomEpoch: 3, sessionEpoch: 4 }),
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "roomStateChanged",
    });
    stateMachine.apply({
      eventSeq: 6,
      room: roomView({ eventSeq: 6, roomEpoch: 3, sessionEpoch: 4 }),
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "roomStateChanged",
    });

    expect(stateMachine.state.latestEventSeq).toBe(7);
  });

  test("stale input frames are ignored by epoch", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 5,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 5, roomEpoch: 3, sessionEpoch: 4 }),
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    stateMachine.apply({
      input: {
        frame: 16,
        payload: [1],
        playerIndex: 1,
      },
      roomEpoch: 2,
      sessionEpoch: 4,
      type: "inputFrame",
    });
    expect(stateMachine.frameClock.snapshot().peerReadFrame).toBeNull();

    stateMachine.apply({
      input: {
        frame: 17,
        payload: [1],
        playerIndex: 1,
      },
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "inputFrame",
    });
    expect(stateMachine.frameClock.snapshot().peerReadFrame).toBe(17);
  });

  test("snapshot runtime messages require the exact active epoch", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 5,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 5, roomEpoch: 3, sessionEpoch: 4 }),
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });

    expect(stateMachine.isRuntimeMessageCurrent({
      chunk: { bytes: [1], index: 0 },
      roomEpoch: 2,
      sessionEpoch: 4,
      type: "snapshotChunk",
    })).toBe(false);
    expect(stateMachine.isRuntimeMessageCurrent({
      manifest: { sha256: "a".repeat(64), totalBytes: 1 },
      roomEpoch: 3,
      sessionEpoch: 5,
      type: "snapshotComplete",
    })).toBe(false);
    expect(stateMachine.isRuntimeMessageCurrent({
      chunk: { bytes: [1], index: 0 },
      roomEpoch: 3,
      sessionEpoch: 4,
      type: "snapshotChunk",
    })).toBe(true);
  });

  test("diagnostics exposes effective pause and frame health", () => {
    const stateMachine = new RoomStateMachine();
    stateMachine.apply({
      eventSeq: 1,
      inputSocketToken: "input-token",
      resumeToken: "token",
      room: roomView({ eventSeq: 1, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "roomJoined",
      yourPlayerIndex: 0,
    });
    stateMachine.apply({
      eventSeq: 2,
      pause: {
        acknowledgedPlayerIndexes: [],
        holders: [],
        pauseAtFrame: 50,
        pausedAtFrame: null,
        reason: "menu",
        requestedByPlayerIndex: 1,
        sequence: 1,
        state: "pausing",
      },
      room: roomView({ eventSeq: 2, roomEpoch: 2, sessionEpoch: 3 }),
      roomEpoch: 2,
      sessionEpoch: 3,
      type: "sessionPauseScheduled",
    });

    expect(stateMachine.diagnostics(1_000)).toMatchObject({
      assignedPlayerIndex: 0,
      effectivePauseReason: "peer",
      heartbeat: "fresh",
      reconnectTicketAvailable: true,
    });
  });
});

File: sdk/kotlin/src/test/kotlin/app/shadowboy/netplay/sdk/state/RoomStateMachineTest.kt
package app.shadowboy.netplay.sdk.state

import app.shadowboy.netplay.sdk.json.NetplayJson
import app.shadowboy.netplay.sdk.protocol.ClientRuntimeState
import app.shadowboy.netplay.sdk.protocol.InputFrame
import app.shadowboy.netplay.sdk.protocol.NearbyStateHashMatchView
import app.shadowboy.netplay.sdk.protocol.PlayerStateHashView
import app.shadowboy.netplay.sdk.protocol.RoomStatus
import app.shadowboy.netplay.sdk.protocol.ServerMessage
import app.shadowboy.netplay.sdk.protocol.ServerFrameRelease
import app.shadowboy.netplay.sdk.protocol.SessionPauseReason
import app.shadowboy.netplay.sdk.protocol.SessionPauseState
import app.shadowboy.netplay.sdk.protocol.SessionPauseView
import app.shadowboy.netplay.sdk.protocol.SnapshotManifest
import app.shadowboy.netplay.sdk.protocol.StateHashMismatchView
import app.shadowboy.netplay.sdk.protocol.StateRecoveryPhase
import app.shadowboy.netplay.sdk.protocol.StateRecoveryView
import app.shadowboy.netplay.sdk.protocol.roomJson
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.seconds

class RoomStateMachineTest {
    @Test
    fun `room joined stores reconnect ticket and assigned player`() {
        val stateMachine = RoomStateMachine()

        val state = stateMachine.apply(
            ServerMessage.RoomJoined(
                eventSeq = 1,
                roomEpoch = 2,
                sessionEpoch = 3,
                yourPlayerIndex = 0,
                resumeToken = "token",
                inputSocketToken = "input-token",
                room = room(status = "waitingForGuest", eventSeq = 1, roomEpoch = 2, sessionEpoch = 3),
            ),
        )

        assertEquals(0, state.assignedPlayerIndex)
        assertEquals(2, stateMachine.reconnectTokens.current()?.roomEpoch)
        assertEquals("token", stateMachine.reconnectTokens.current()?.resumeToken)
    }

    @Test
    fun `recovery resync updates state and keeps assignment`() {
        val stateMachine = RoomStateMachine()
        stateMachine.apply(
            ServerMessage.RoomJoined(
                eventSeq = 1,
                roomEpoch = 2,
                sessionEpoch = 3,
                yourPlayerIndex = 0,
                resumeToken = "token",
                inputSocketToken = "input-token",
                room = room(status = "waitingForGuest", eventSeq = 1, roomEpoch = 2, sessionEpoch = 3),
            ),
        )

        val state = stateMachine.apply(
            ServerMessage.RecoveryResyncRequired(
                eventSeq = 10,
                roomEpoch = 5,
                sessionEpoch = 9,
                room = room(status = "checkingCompatibility", eventSeq = 10, roomEpoch = 5, sessionEpoch = 9),
            ),
        )

        assertEquals(0, state.assignedPlayerIndex)
        assertEquals(10, state.latestEventSeq)
        assertEquals(RoomStatus.CheckingCompatibility, state.room?.status)
        assertEquals(NetplayResyncReason.Recovery, state.resync?.reason)
        assertEquals(NetplayResyncPhase.Requested, state.resync?.phase)
        assertEquals(NetplayResyncRole.Host, state.resync?.role)
        assertEquals(true, state.resync?.mustSendSnapshot)
        assertEquals(10, state.resync?.eventSeq)
        assertEquals(5, state.resync?.roomEpoch)
        assertEquals(9, state.resync?.sessionEpoch)
        assertEquals(true, stateMachine.resync.shouldPauseEmulation())
        assertEquals(true, stateMachine.resync.shouldClearPredictionBuffers())
        assertEquals(5, stateMachine.reconnectTokens.current()?.roomEpoch)
    }

    @Test
    fun `state hash mismatch enters resync until session restarts`() {
        val stateMachine = RoomStateMachine()
        stateMachine.apply(
            ServerMessage.RoomJoined(
                eventSeq = 1,
                roomEpoch = 2,
                sessionEpoch = 3,
                yourPlayerIndex = 0,
                resumeToken = "token",
                inputSocketToken = "input-token",
                room = room(status = "waitingForGuest", eventSeq = 1, roomEpoch = 2, sessionEpoch = 3),
            ),
        )

        val mismatch = StateHashMismatchView(
            frame = 120,
            repairFrame = 120,
            hashes = listOf(
                PlayerStateHashView(playerIndex = 0, sha256 = "a".repeat(64)),
                PlayerStateHashView(playerIndex = 1, sha256 = "b".repeat(64)),
            ),
            nearbyMatches = emptyList<NearbyStateHashMatchView>(),
        )
        val resyncing = stateMachine.apply(
            ServerMessage.StateHashMismatch(
                eventSeq = 11,
                roomEpoch = 2,
                sessionEpoch = 4,
                mismatch = mismatch,
                room = room(status = "checkingCompatibility", eventSeq = 11, roomEpoch = 2, sessionEpoch = 4),
            ),
        )

        assertEquals(NetplayResyncReason.StateHashMismatch, resyncing.resync?.reason)
        assertEquals(NetplayResyncPhase.Requested, resyncing.resync?.phase)
        assertEquals(mismatch, resyncing.resync?.mismatch)
        assertEquals(11, resyncing.resync?.eventSeq)

        val started = stateMachine.apply(
            ServerMessage.StartSession(
                eventSeq = 12,
                roomEpoch = 2,
                sessionEpoch = 4,
                startFrame = 0,
                room = room(status = "playing", eventSeq = 12, roomEpoch = 2, sessionEpoch = 4),
            ),
        )

        assertNull(started.resync)
    }

    @Test
    fun `v5 recovery hydrates and resets runtime only once per recovery id`() {
        val stateMachine = RoomStateMachine()
        val preparing = recovery(StateRecoveryPhase.Preparing)
        val preparingRoom = room(
            status = "repairingState",
            eventSeq = 1,
            roomEpoch = 2,
            sessionEpoch = 3,
        ).copy(
            protocol = room(
                status = "repairingState",
                eventSeq = 1,
                roomEpoch = 2,
                sessionEpoch = 3,
            ).protocol.copy(protocolVersion = 5),
            stateRecovery = preparing,
        )

        val hydrated = stateMachine.apply(
            ServerMessage.RoomJoined(
                eventSeq = 1,
                roomEpoch = 2,
                sessionEpoch = 3,
                yourPlayerIndex = 0,
                resumeToken = "token",
                inputSocketToken = "input-token",
                room = preparingRoom,
            ),
        )

        assertEquals(preparing, hydrated.stateRecovery)
        assertNull(hydrated.resync)
        assertTrue(hydrated.runtimeResetRequired)
        assertEquals(NetplayEffectivePauseReason.StateResync, stateMachine.effectivePauseReason())

        stateMachine.acknowledgeRuntimeReset()
        stateMachine.apply(
            ServerMessage.StateRecoveryPrepare(
                eventSeq = 2,
                roomEpoch = 2,
                sessionEpoch = 3,
                recovery = preparing,
                room = preparingRoom.copy(eventSeq = 2),
            ),
        )
        assertFalse(stateMachine.state.runtimeResetRequired)

        val manifest = SnapshotManifest(
            snapshotId = "recovery-7",
            repairFrame = 600,
            totalBytes = 4,
            sha256 = "c".repeat(64),
        )
        val committed = recovery(StateRecoveryPhase.Committed, manifest)
        val committedRoom = preparingRoom.copy(
            eventSeq = 3,
            sessionEpoch = 4,
            status = RoomStatus.CheckingCompatibility,
            stateRecovery = committed,
        )
        stateMachine.apply(
            ServerMessage.StateRecoveryCommitted(
                eventSeq = 3,
                roomEpoch = 2,
                sessionEpoch = 4,
                recovery = committed,
                room = committedRoom,
            ),
        )

        assertEquals(committed, stateMachine.state.stateRecovery)
        assertFalse(stateMachine.state.runtimeResetRequired)

        stateMachine.apply(
            ServerMessage.StartSession(
                eventSeq = 4,
                roomEpoch = 2,
                sessionEpoch = 4,
                startFrame = 600,
                room = committedRoom.copy(eventSeq = 4, status = RoomStatus.StartScheduled),
            ),
        )
        assertEquals(committed, stateMachine.state.stateRecovery)

        stateMachine.markStateRecoveryFrameExecuted(2, 4, 601)
        assertEquals(committed, stateMachine.state.stateRecovery)
        stateMachine.markStateRecoveryFrameExecuted(2, 4, 600)
        assertNull(stateMachine.state.stateRecovery)
        assertNull(stateMachine.state.room?.stateRecovery)
    }

    @Test
    fun `resync coordinator exposes snapshot decisions`() {
        val stateMachine = RoomStateMachine()
        stateMachine.apply(
            ServerMessage.RoomJoined(
                eventSeq = 1,
                roomEpoch = 2,
                sessionEpoch = 3,
                yourPlayerIndex = 0,
                resumeToken = "token",
                inputSocketToken = "input-token",
                room = room(status = "waitingForGuest", eventSeq = 1, roomEpoch = 2, sessionEpoch = 3),
            ),
        )

        stateMachine.apply(
            ServerMessage.StateHashMismatch(
                eventSeq = 11,
                roomEpoch = 2,
                sessionEpoch = 4,
                mismatch = StateHashMismatchView(
                    frame = 120,
                    repairFrame = 120,
                    hashes = listOf(
                        PlayerStateHashView(playerIndex = 0, sha256 = "a".repeat(64)),
                        PlayerStateHashView(playerIndex = 1, sha256 = "b".repeat(64)),
                    ),
                    nearbyMatches = emptyList<NearbyStateHashMatchView>(),
                ),
                room = room(status = "checkingCompatibility", eventSeq = 11, roomEpoch = 2, sessionEpoch = 4),
            ),
        )

        stateMachine.resync.markSnapshotNeeded(nowMs = 1_000)

        assertEquals(true, stateMachine.resync.shouldSendHostSnapshot())
        assertEquals(false, stateMachine.resync.shouldWaitForSnapshot())
    }

    @Test
    fun `heartbeat tracker reports stale and recovery timeout`() {
        val tracker = HeartbeatTracker(
            HeartbeatPolicy(staleAfter = 5.seconds, recoveryAfter = 10.seconds),
        )
        tracker.markAck(ServerMessage.HeartbeatAck(1, 2, 3), nowMillis = 1_000)

        assertEquals(HeartbeatHealth.Fresh, tracker.health(nowMillis = 1_500))
        assertEquals(HeartbeatHealth.Stale, tracker.health(nowMillis = 6_000))
        assertEquals(HeartbeatHealth.RecoveryTimedOut, tracker.health(nowMillis = 11_000))
        assertEquals(
            ClientRuntimeState.Playing,
            tracker.heartbeatMessage(2, 3, 1, 24, ClientRuntimeState.Playing).runtimeState,
        )
    }

    @Test
    fun `heartbeat can consume runtime telemetry safely`() {
        val tracker = HeartbeatTracker()
        val telemetry = RuntimeTelemetryTracker()

        telemetry.markLocalFrame(90)
        telemetry.recordRoundTrip(40)
        telemetry.recordRoundTrip(48)
        telemetry.recordStall()
        telemetry.recordCatchUpFrames(2)

        val heartbeat = tracker.heartbeatMessage(
            roomEpoch = 2,
            sessionEpoch = 3,
            latestEventSeq = 4,
            localFrame = null,
            runtimeState = ClientRuntimeState.Playing,
            telemetry = telemetry,
        )

        assertEquals(90, heartbeat.localFrame)
        assertEquals(48, heartbeat.network?.roundTripMs)
        assertEquals(1, heartbeat.network?.stallCount)
        assertEquals(2, heartbeat.network?.catchUpFrames)
        assertEquals(0, telemetry.snapshot().network.stallCount)
    }

    @Test
    fun `state hash reporter normalizes hashes and deduplicates frames`() {
        val reporter = StateHashReporter(StateHashReporterPolicy(reportEveryFrames = 30))

        assertEquals(true, reporter.shouldReport(0))
        reporter.stateHashMessage(
            roomEpoch = 2,
            sessionEpoch = 3,
            frame = 0,
            sha256 = "a".repeat(64),
        )
        assertEquals(false, reporter.shouldReport(29))
        assertEquals(true, reporter.shouldReport(30))
        assertEquals(
            "a".repeat(64),
            reporter.stateHashMessage(
                roomEpoch = 2,
                sessionEpoch = 3,
                frame = 30,
                sha256 = "A".repeat(64),
            ).report.sha256,
        )
        assertEquals(false, reporter.shouldReport(30))
        assertEquals(false, reporter.shouldReport(59))
        assertEquals(true, reporter.shouldReport(61))
    }

    @Test
    fun `pause coordinator creates and clears pause messages`() {
        val pause = PauseCoordinator()
        pause.apply(
            SessionPauseView(
                sequence = 3,
                state = SessionPauseState.Pausing,
                reason = SessionPauseReason.Menu,
                requestedByPlayerIndex = 0,
                pauseAtFrame = 120,
                pausedAtFrame = null,
                acknowledgedPlayerIndexes = emptyList(),
                holders = emptyList(),
            ),
        )

        assertEquals(3, pause.pauseReached(4, 5, pausedAtFrame = 121).sequence)
        assertEquals(3, pause.requestResume(4, 5, SessionPauseReason.Menu).sequence)

        pause.clear(3)
        assertNull(pause.currentPause)
    }

    @Test
    fun `frame clock tracks server frame and peer read frame`() {
        val stateMachine = RoomStateMachine()

        stateMachine.apply(
            ServerMessage.ServerFrameMessage(
                ServerFrameRelease(
                    roomEpoch = 1,
                    sessionEpoch = 1,
                    frame = 18,
                    canonicalFrame = 20,
                ),
            ),
        )
        stateMachine.apply(
            ServerMessage.InputFrameMessage(
                roomEpoch = 1,
                sessionEpoch = 1,
                input = InputFrame(
                    playerIndex = 1,
                    frame = 16,
                    payload = listOf(1),
                ),
            ),
        )
        stateMachine.frameClock.markLocalFrame(80)

        val diagnostics = stateMachine.frameClock.snapshot()
        assertEquals(20, diagnostics.canonicalFrame)
        assertEquals(18, diagnostics.serverFrame)
        assertEquals(16, diagnostics.peerReadFrame)
        assertEquals(true, diagnostics.stalled)
    }

    private fun room(
        status: String,
        eventSeq: Long,
        roomEpoch: Long,
        sessionEpoch: Long,
    ) =
        NetplayJson.format.decodeFromString(
            app.shadowboy.netplay.sdk.protocol.RoomView.serializer(),
            roomJson(
                status = status,
                eventSeq = eventSeq,
                roomEpoch = roomEpoch,
                sessionEpoch = sessionEpoch,
            ),
        )

    private fun recovery(
        phase: StateRecoveryPhase,
        manifest: SnapshotManifest? = null,
    ) =
        StateRecoveryView(
            recoveryId = 7,
            phase = phase,
            repairFrame = 600,
            mismatch = StateHashMismatchView(
                frame = 600,
                repairFrame = 600,
                hashes = listOf(
                    PlayerStateHashView(playerIndex = 0, sha256 = "a".repeat(64)),
                    PlayerStateHashView(playerIndex = 1, sha256 = "b".repeat(64)),
                ),
            ),
            pinnedSnapshot = manifest,
        )

}
